File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/monitoring/adapter/MongoMetricsRollupRepository.java
package com.study.webflux.rag.infrastructure.monitoring.adapter;

import java.time.Instant;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Repository;

import com.study.webflux.rag.domain.monitoring.model.MetricsGranularity;
import com.study.webflux.rag.domain.monitoring.model.MetricsRollup;
import com.study.webflux.rag.domain.monitoring.port.MetricsRollupRepository;
import com.study.webflux.rag.infrastructure.monitoring.document.MetricsRollupDocument;
import com.study.webflux.rag.infrastructure.monitoring.repository.SpringDataMetricsRollupRepository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Repository
@RequiredArgsConstructor
public class MongoMetricsRollupRepository implements MetricsRollupRepository {

	private final SpringDataMetricsRollupRepository repository;

	@Override
	public Mono<MetricsRollup> save(MetricsRollup rollup) {
		return Mono.just(rollup)
			.map(MetricsRollupDocument::fromDomain)
			.flatMap(repository::save)
			.map(MetricsRollupDocument::toDomain);
	}

	@Override
	public Flux<MetricsRollup> findByGranularityAndBucketStartBetweenOrderByBucketStartAsc(
		MetricsGranularity granularity,
		Instant startTime,
		Instant endTime) {
		return repository.findByGranularityAndBucketStartBetweenOrderByBucketStartAsc(
			granularity.name(),
			startTime,
			endTime)
			.map(MetricsRollupDocument::toDomain);
	}
}

File: infrastructure/src/main/kotlin/com/miyou/app/infrastructure/monitoring/adapter/MongoMetricsRollupRepository.kt
package com.miyou.app.infrastructure.monitoring.adapter

import com.miyou.app.domain.monitoring.model.MetricsGranularity
import com.miyou.app.domain.monitoring.model.MetricsRollup
import com.miyou.app.domain.monitoring.port.MetricsRollupRepository
import com.miyou.app.infrastructure.monitoring.document.MetricsRollupDocument
import com.miyou.app.infrastructure.monitoring.repository.SpringDataMetricsRollupRepository
import org.springframework.stereotype.Repository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Instant

@Repository
class MongoMetricsRollupRepository(
    private val repository: SpringDataMetricsRollupRepository,
) : MetricsRollupRepository {
    override fun save(rollup: MetricsRollup): Mono<MetricsRollup> =
        Mono
            .just(rollup)
            .map(MetricsRollupDocument::fromDomain)
            .flatMap(repository::save)
            .map(MetricsRollupDocument::toDomain)

    override fun findByGranularityAndBucketStartBetweenOrderByBucketStartAsc(
        granularity: MetricsGranularity,
        startTime: Instant,
        endTime: Instant,
    ): Flux<MetricsRollup> =
        repository
            .findByGranularityAndBucketStartBetweenOrderByBucketStartAsc(
                granularity.name,
                startTime,
                endTime,
            ).map(MetricsRollupDocument::toDomain)
}
